contracts/RemmeBridge.sol

```diff
@@ -1,10 +1,240 @@
+            keyHolder = msg.sender;
+        }
+
+	    AtomicSwap memory swap = AtomicSwap({
+            swapId: _swapId,
+            senderAddress: msg.sender,
+            receiverAddress: _receiverAddress,
+            keyHolderAddress: keyHolder,
+            remchainAddress: _remchainAddress,
+            amount: amountToSwap,
+            emailAddressEncryptedOptional: _emailAddressEncryptedOptional,
+            secretLock: _secretLock,
+            secretKey: "", //will be set only on closing swap
+            timelock: lock
+            });
+        swaps[swap.swapId] = swap;
+
+	    require(REMToken.transferFrom(msg.sender, address(this), amountToSwap));
+
+        swapStates[swap.swapId] = State.OPENED;
+        emit OpenSwap(swap.swapId);
+    }
+
+	/* @notice should be called after timelock
+	*/
+    function expireSwap(bytes32 _swapId) onlyOverdueSwap(_swapId) external {
+
+        require(msg.sender == swaps[_swapId].senderAddress);
+        require(swapStates[_swapId] == State.OPENED || swapStates[_swapId] == State.APPROVED);
+
+        AtomicSwap storage currentSwap = swaps[_swapId];
+        REMToken.transfer(msg.sender, currentSwap.amount);
+
+        swapStates[_swapId] = State.EXPIRED;
+        emit ExpireSwap(_swapId);
+    }
+
+    /* @notice should be called only by Bob after opening swap by Alice, otherwise
+    *  secretLock was set in openSwap()
+    *  @param _secretLock secretKey hashed by Bob
+    */
+    function setSecretLock(bytes32 _swapId, bytes32 _secretLock)
+    onlyNotOverdueSwap(_swapId)
+    onlyOpenedState(_swapId)
+    external {
+
+        require(swaps[_swapId].secretLock == bytes32(0));
+        require(msg.sender == swaps[_swapId].keyHolderAddress);
+        require(_secretLock != bytes32(0));
+
+        swaps[_swapId].secretLock = _secretLock;
+        emit SetSecretLock(_swapId);
+    }
+
+    /* @notice should be called only by Alice after setting lock
+    */
+    function approveSwap(bytes32 _swapId) onlyNotOverdueSwap(_swapId) onlyOpenedState(_swapId) external payable {
+
+        require(swaps[_swapId].secretLock != bytes32(0));
+        require(msg.sender == swaps[_swapId].senderAddress);
+
+        swapStates[_swapId] = State.APPROVED;
+        emit ApproveSwap(_swapId);
+    }
+
+	/* @notice should be called with provided secret key
+    */
+    function closeSwap(bytes32 _swapId, bytes32 _secretKey) onlyNotOverdueSwap(_swapId) external {
+
+        require(swaps[_swapId].secretLock != bytes32(0));
+        require(msg.sender == swaps[_swapId].receiverAddress);
+
+	    if (swaps[_swapId].senderAddress == swaps[_swapId].keyHolderAddress) {
+		    require(swapStates[_swapId] == State.OPENED);
+	    } else {
+		    require(swapStates[_swapId] == State.APPROVED);
+        }
+        require(keccak256(_secretKey) == swaps[_swapId].secretLock);
+
+	    AtomicSwap storage currentSwap = swaps[_swapId];
+	    if (currentSwap.receiverAddress == atomicSwapProvider) {
```

Comment:
Why do we care about atomicSwapProvider if we make a universal contract where a receiver can be either bot or any other agent?

Reply:
provider needed for calculating and transferring fee automatically by contract

Reply:
if it will be performed offchain, there might be stuck swaps, when fee will be changed and provider don't allow insufficient amount for swap